### ISSUE TITLE
ModuleNotFoundError when pytest not in venv

### ISSUE THREAD

**Issue by mkesper** [pytest-dev/pytest#11111]
If pytest is installed system-wide but not in a venv, pytest will run but not find the dependencies installed in the venv. It would be nice if pytest would trigger a descriptive error if the ENV var VIRTUAL_ENV is set but pytest is not running from inside this venv.

pytest 7.2.1 from Debian bookworm (12)

Reproduce: 

1. Ensure pytest is installed system-wide
2. Create venv
3. Activate venv and install any module (e.g. foo) not present in system (but not pytest)
4. Create a test file importing this module
5. Run pytest, it will produce "E   ModuleNotFoundError: No module named 'foo'"

**Comment by RonnyPfannschmidt** [comment]
A warning would be nice

**Comment by symonk** [comment]
just to note I don't think `VIRTUAL_ENV` is reliable here as its only set on activation, `sys.prefix` / `sys.base_prefix` are possibly a more reliable method for detection?

**Comment by The-Compiler** [comment]
But that's the whole point of the warning, no? To warn you if you have a virtualenv activated and are using pytest from outside it.

**Comment by symonk** [comment]
Yes sorry, I'm wondering if it was only set for the core `venv` module activation and there might be other use cases that people can use for venvs, poetry, pyenv etc (not sure if thats a reliable way to cover all edge cases basically)

**Comment by The-Compiler** [comment]
Probably not I suppose... but then again, a warning might already be useful if it only catches the most common case.

**Comment by mkesper** [comment]
@The-Compiler The warning should tell I'm using pytest from outside of the virtualenv, not breaking on some random module import. Or was that what you meant?

**Comment by The-Compiler** [comment]
@mkesper I'm confused by the question. What I'm saying is that checking `sys.prefix` and `sys.base_prefix` doesn't make much sense here, because the whole point is that you're not running the virtualenv pytest in the first place.

**Comment by mkesper** [comment]
@The-Compiler I agree 100%. :)

**Comment by The-Compiler** [comment]
Actually, thinking about it a bit more, I'm not sure how this can even happen.

Activating a virtualenv sets *both* `VIRTUAL_ENV` *and* `PATH`, and installing `pytest` in the virtualenv means that there should be a `pytest` binary in its bin (which is in your path).

So the only way I can see this warning trigger is if you manually mess with `PATH` after activating the virtualenv, which seems like "you get to keep both pieces" territory for me.

I tried to reproduce your steps in a `debian:12` docker image, and things are working as expected:

```
root@73d3b126cc6c:/# apt update
...
root@73d3b126cc6c:/# apt install -q python3 python3-venv python3-pytest
...
root@73d3b126cc6c:/# pytest --version
pytest 7.2.1
root@73d3b126cc6c:/# which pytest
/usr/bin/pytest
root@73d3b126cc6c:/# python3 -c "import requests"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'requests'
root@73d3b126cc6c:/# python3 -m venv .venv
root@73d3b126cc6c:/# source .venv/bin/activate                      
(.venv) root@73d3b126cc6c:/# pip install -q pytest requests
(.venv) root@73d3b126cc6c:/# echo "import requests" > test_x.py
(.venv) root@73d3b126cc6c:/# which pytest
/.venv/bin/pytest
(.venv) root@73d3b126cc6c:/# pytest test_x.py
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-7.4.0, pluggy-1.2.0
rootdir: /
collected 0 items                                                              

============================ no tests ran in 0.04s =============================
```

So, @mkesper, can you clarify how this could actually happen in practice, i.e. what scenario exactly such a warning would trigger in?

**Comment by The-Compiler** [comment]
Ah. I spoke to soon, it happens when you *don't* install `pytest` in the virtual environment (i.e. only `pip install -q requests` above):

```console
(.venv) root@26883eaabb26:/# which pytest
/usr/bin/pytest
(.venv) root@26883eaabb26:/# pytest test_x.py 
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-7.2.1, pluggy-1.0.0+repack
rootdir: /
collected 0 items / 1 error                                                    

==================================== ERRORS ====================================
__________________________ ERROR collecting test_x.py __________________________
ImportError while importing test module '/test_x.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
usr/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_x.py:1: in <module>
    import requests
E   ModuleNotFoundError: No module named 'requests'
=========================== short test summary info ============================
ERROR test_x.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.03s ===============================
```

In which case indeed showing a warning when `VIRTUAL_ENV` is set but we're not running in that virtualenv (*that* part should probably be based on `sys.prefix`) might make sense.

Not sure if it should *always* be showh, or only shown if we have an `ImportError` during collection or somesuch.

**Comment by RonnyPfannschmidt** [comment]
Should always show

**Comment by mkesper** [comment]
@The-Compiler thanks for re-testing. Yes, that's exactly where I think an error would be nice.
